server/e2e_test.go

```diff
@@ -407,35 +405,54 @@ func (h *testHarness) AssertBroadcastPhase1Announcement(all []*testClient) {
 	}
 }
 
-// AssertBroadcastBlame confirms and consumes the blame broadcast to all
+// WaitBroadcastBlame confirms and consumes the blame broadcast to all
 // pool players.
-func (h *testHarness) AssertBroadcastBlame(expected *message.Signed, pool []*testClient) {
+func (h *testHarness) WaitBroadcastBlame(expected *message.Signed, pool []*testClient) {
 	for _, c := range pool {
 		actualPacket, err := c.inbox.PopOldest()
 		if err != nil {
-			panic(err)
+			h.t.Fatal(err)
 		}
 		assert.Len(h.t, actualPacket.message.GetPacket(), 1)
 		actualMsg := actualPacket.message.GetPacket()[0]
 		assert.Equal(h.t, expected, actualMsg)
 	}
 }
 
-// AssertNotConnected confirms that the client is not connected to a server
-func (h *testHarness) AssertNotConnected(c *testClient) {
+// WaitNotConnected confirms that both sides of the client connection are closed
+// and that client's connection is not in the server's lookup table
+func (h *testHarness) WaitNotConnected(c *testClient) {
 	err := retry.Do(
 		func() error {
-			_, err := c.conn.Read([]byte{})
-			if err == io.EOF {
-				return nil
+			h.tracker.mutex.RLock()
+			defer h.tracker.mutex.RUnlock()
+			// confirm removed from connection lookup
+			_, isTracked := h.tracker.connections[c.remoteConn]
+			if isTracked {
+				return fmt.Errorf("server thinks client is still connected")
+			}
+
+			// confirm both sides of connection are closed
+			if c.conn != nil {
+				_, err := c.conn.Read([]byte{})
+				if (err != io.EOF) && (err != io.ErrClosedPipe) {
+					return fmt.Errorf("client side of connection still active")
+				}
+			}
+			if c.remoteConn != nil {
+				_, err := c.remoteConn.Read([]byte{})
+				if (err != io.EOF) && (err != io.ErrClosedPipe) {
+					return fmt.Errorf("server side of connection still active")
+				}
 			}
-			return fmt.Errorf("still connected")
+			// all evidence says client is disconnected
+			return nil
```

Comment:
This was the cause of a race condition that the race detector could not find. The previous code was assuming that as soon as a disconnect was confirmed at the connection level, that the server had also acted on the disconnection. Now it confirms the disconnect at the connection level and also makes sure that the server is not still holding the connection in the lookup table.

Reply:
Nice find!